### PR TITLE
Supply Stash

### PR DESCRIPTION
**Pull request by vinioliveirasilva**
Fix open stash
Fix add stackable item in stash
fix multilayer stow (add a container)

TODO:
Add stow container from Menu Context (server must support protocol 12+), so, it will be implemented later
Validation 
- [x]  player has cap to withdraw 
- [x]  player has container to withdraw 
- [x]  player has stash capacity to add a new item 
- [x]  Stash Retrieve from manager loot container is defined (need quickloot in cpp to do that)

Add migration

### CODE REVIEW

**Comment by marcosvf132** [comment]
FIXED `Is there a way to convert item cid to item id? need this to add withdraw function. `
Also, is crashing server when moving items to the floor/depot chest.

**Comment by marcosvf132** [comment]
> Fix open stash
> Fix add stackable item in stash
> TODO:
> Add multilayer stow (add a container)
> Add withdraw
> Add migration

FIX 'add withdraw' with the following changes:

- On **\otservbr-global\src\game.cpp**
line 1078:
```lua
uint32_t flags = 0;
	int32_t index = toIndex;
	Item* toItem = nullptr;
	Cylinder* subCylinder = toCylinder->queryDestination(index, *item, &toItem, flags);
	if (subCylinder->getItem()->getID() == ITEM_SUPPLY_STASH) { //should be here?
		if (!(item->isStackable()) || 
			item->getID() == ITEM_GOLD_COIN ||
			item->getID() == ITEM_PLATINUM_COIN ||
			item->getID() == ITEM_CRYSTAL_COIN ) {
			return player->sendCancelMessage(RETURNVALUE_NOTPOSSIBLE);
		}
		std::ostringstream query;
		uint32_t itemCount = count;
		query << "SELECT * FROM `player_stash` WHERE `player_id` = ";
		query << player->getGUID() << " AND `item_id` = " << item->getClientID() << ";";
		DBResult_ptr hasItem = Database::getInstance().storeQuery(query.str());
		query.str("");
		if (!hasItem) {
			query << "INSERT INTO `player_stash` (`player_id`,`item_id`,`item_id2`, `item_count`) VALUES (";
			query << player->getGUID() << ", ";
			query << item->getClientID() << ", ";
			query << item->getID() << ", ";
			query << itemCount << " )";
		} else {
			int32_t qntInStash = hasItem->getNumber<int32_t>("item_count");
			query << "UPDATE `player_stash` SET `item_count` = ";
			query << itemCount + qntInStash;
			query << " WHERE `player_id` = " << player->getGUID();
			query << " AND `item_id2` = " << item->getID();
			query << " AND `item_id` = " << item->getClientID() << ";";
		}
		bool result = Database::getInstance().executeQuery(query.str());
		query.str("");
		if (result) {
			internalRemoveItem(item, itemCount);
			query << "Stowed " << itemCount << " object" << (itemCount > 1 ? "s." : ".");
			return player->sendCancelMessage(query.str());
		}
		return player->sendCancelMessage(RETURNVALUE_NOTPOSSIBLE);
	}
```
- On **\otservbr-global\data\globalevents\scripts\startup.lua**

```lua
...
installstashdb() -- At any place
...
-- At the end
function installstashdb()
	db.query([[CREATE TABLE IF NOT EXISTS `player_stash` (
  `player_id` INT(16) NOT NULL,
  `item_id` INT(32) NOT NULL,
  `item_id2` INT(32) NOT NULL,
  `item_count` INT(32) NOT NULL) ENGINE=InnoDB DEFAULT CHARSET=utf8;
   ]])
end
```
- On **\otservbr-global\data\modules\scripts\supplystash\supplystash.lua**

```lua
...
	elseif action == 3 then -- Withdraw from stash window
		local testando = msg:getU16()
		local itemCount = msg:getU32()
		print('withdraw')		
	local boid = db.storeQuery("SELECT `item_id2` FROM `player_stash` WHERE `player_id` = " .. player:getGuid() .. " AND `item_id` = " .. testando)
    local boidstamp = result.getStream(boid, "item_id2")
    result.free(boid)
		player:addItem(math.floor(boidstamp),itemCount)
	local boname = db.storeQuery("SELECT `item_count` FROM `player_stash` WHERE `player_id` = " .. player:getGuid() .. " AND `item_id` = " .. testando)
    local bonamestammp = result.getStream(boname, "item_count")
    result.free(boname)
	local ficouQT = math.floor(bonamestammp - itemCount)
		if ficouQT  == 0 then
		db.asyncQuery('DELETE FROM `player_stash` WHERE `item_id` = ' .. testando)
		else
		db.asyncQuery('UPDATE `player_stash` SET `item_count` = ' .. ficouQT .. ' WHERE `item_id` = '.. testando ..' AND `player_id` = ' .. player:getGuid())
		end
	end
end
-- End of archive
```
With this you already can withdraw items from the stash, but the problem is that when you withdraw, you need to close and open again the stashwindow to refresh the item count, if you dont do so, the item will get negative value and will bug your player_stash database, then you will need to delete the player_stash label on your database.

**Need to find a way to force a refresh.**

TODO: 
Add refresh stash counter
Add multilayer stow (add a container)
Add migration (?)

**Comment by vinioliveirasilva** [comment]
> FIXED `Is there a way to convert item cid to item id? need this to add withdraw function. `
> Also, is crashing server when moving items to the floor/depot chest.

Yeh, using Game namespace: like this

(Game.getItemIdByClientId(ItemClientId)):getId() and we get item id

**Comment by marcosvf132** [comment]
> FIXED Is there a way to convert item cid to item id? need this to add withdraw function. 
> Also, is crashing server when moving items to the floor/depot chest.
> 
> Yeh, using Game namespace: like this
> (Game.getItemIdByClientId(ItemClientId)):getId() and we get item id

Is there a way to do the oposite? itemID to ClientID? i need that to make stow container work.

**Comment by FakeShinoda** [comment]
This system can be easily wrote all in c++, we have alot of systems in modules that need to be rewrite in source.
Btw, good initiative !

**Comment by marcosvf132** [comment]
Hello again, i have done a script that you can 'stow' your main container. But it only work when you try to stow you main backpack when you select 'Stow container's contant' on the main backpack menu.

Other problem is that when you try to stow you main backpack, it dont stow all you items at once, on my tests, i tryed to stow 6k items, but when i click on it, it only stow almost half of the items at once, so you need to click on the menu more than once.  need to test it more because sometimes you lose some items on the process. in my example when i stow 6k items, i lost 100 items at the end.

I will post it here because maybe you can figure out how to fix it. 
**i'm not a pro, just trying to contribute with you guys.**

-  On \otservbr-global\data\modules\scripts\supplystash\supplystash.lua
```lua
...



	elseif action == 1 then
		local test11 = msg:getByte() -- 255 (?)
		local test12 = msg:getByte() -- 255 (?)
		local test13 = msg:getByte() -- SLOT ID when stow from main backpack, maybe it is the container position ?
		local myItem = player:getSlotItem(test13) -- Main backpack position (when try to stow from main backpack)
        searchitemsStash(myItem, player)
		print('stow container')	
	elseif action == 2 then
...
...
function searchitemsStash(test, JOGADOR)
	if test then
		for i = 0, test:getSize() do
				local item = test:getItem(i)
			if item then
					if item:isContainer() then
						print('container')	
						searchitemsStash(item, JOGADOR)
					end
				if (item:getCount() > 1) and not(item:getId() == 2160) and not(item:getId() == 2152) and not(item:getId() == 2148) and not(item:isContainer()) then
						local zero = 0
						local clientType = ItemType(item:getId())
						db.query("INSERT INTO `player_stash` (`player_id`, `item_id`, `item_count`) VALUES (" .. JOGADOR:getGuid() .. ", " .. clientType:getClientId() .. ", " .. zero .. ");")
						local items2 = loadStash(JOGADOR, clientType:getClientId(), 1000000)
					if items2[itemId] ~= nill then
					else
							local resulttestId = db.storeQuery("SELECT `item_count` FROM `player_stash` WHERE `player_id` = " .. JOGADOR:getGuid() .. " AND `item_id` = " .. clientType:getClientId())
						if resulttestId ~= false then
							local testando12 = result.getDataInt(resulttestId , "item_count")
							result.free(resulttestId)
							local TOTAL = math.floor(testando12 + item:getCount())
							db.asyncQuery('UPDATE `player_stash` SET `item_count` = ' .. TOTAL .. ' WHERE `item_id` = '.. clientType:getClientId() ..' AND `player_id` = ' .. JOGADOR:getGuid())
							item:remove(item:getCount())		
						end		
					end
				end	
			end
		end	
	end
end
```
As i said, i'm not a pro on programing, so it need several changes to make it perfect. But i only posted it here so maybe i can help someone to find a way to implement the 'stow' option in game.

File: data/modules/scripts/supplystash/supplystash.lua [hunk 1]
```diff
@@ -13,71 +15,67 @@ SupplyStash.C_Packets = { Withdraw = 0x28 }
 function onRecvbyte(player, msg, byte)
```
Comment (lgrossi) [inline data/modules/scripts/supplystash/supplystash.lua]: That should be in cpp, protocolgame should control the network traffic and then call the necessary callback in lua.

File: data/modules/scripts/supplystash/supplystash.lua [hunk 2]
```diff
@@ -13,71 +15,67 @@ SupplyStash.C_Packets = { Withdraw = 0x28 }
...
 	local depotSlots = 1
+
 	for _, itemCount in pairs(items) do
-		print(math.floor(math.min(itemCount / 100, 1)))
-		depotSlots = depotSlots + math.ceil(math.min(itemCount / 100, 1))
+		depotSlots = depotSlots + math.ceil(math.max(itemCount / 100, 1))
 		count = count + 1
 	end
 
+	local msg = NetworkMessage()
```
Comment (lgrossi) [inline data/modules/scripts/supplystash/supplystash.lua]: Same here.

File: data/modules/scripts/supplystash/supplystash.lua [hunk 3]
```diff
@@ -89,6 +87,18 @@ SupplyStash.parseStash = function(player, msg)
 		print('stow all items of this')
 	elseif action == 3 then -- Withdraw from stash window
 		-- retrieve
-		print('withdraw')
+		local itemId = msg:getU16()
+		local itemCount = msg:getU32()
+		local items = loadStash(player, itemId, itemCount)
```
Comment (lgrossi) [inline data/modules/scripts/supplystash/supplystash.lua]: Those messages operations should also be in cpp

**Review by lgrossi — COMMENTED** [review]
Looks good Im general. However, we should stop handling messages in lua. Messages operations should be centered in cpp, for many reasons: consistency, maintenance, security, etc. Lua is able to do exatcly the same thing via a callback that receive the parsed attributes of the message, not the message itself.

**Comment by marcosvf132** [comment]
STOW CONTAINER: This script below can stow your main backpack, but **only works** if you click on 'stow container's content' on your main backpack menu.

- **On \otservbr-global\data\modules\scripts\supplystash\supplystash.lua**
```lua
...
	elseif action == 1 then
		local test11 = msg:getByte() -- 255 (?)
		local test12 = msg:getByte() -- 255 (?)
		local test13 = msg:getByte() -- SLOT ID when stow from main backpack, maybe it is the container position ?
		local myItem = player:getSlotItem(test13) -- Main backpack position (when try to stow from main backpack)
        searchitemsStash(myItem, player)
		print('stow container')	
	elseif action == 2 then
...
...

function searchitemsStash(test, JOGADOR)
	if test then
		for i = 0, test:getSize() do
				local item = test:getItem(i)
			if item then
					if item:isContainer() then
						searchitemsStash(item, JOGADOR)
					end
				if (item:getCount() > 1) and not(item:getId() == 2160) and not(item:getId() == 2152) and not(item:getId() == 2148) and not(item:isContainer()) then
				local clientType = ItemType(item:getId())
				local resulttestId = db.storeQuery("SELECT `item_count` FROM `player_stash` WHERE `player_id` = " .. JOGADOR:getGuid() .. " AND `item_id` = " .. clientType:getClientId())
				result.free(resulttestId)
				if resulttestId == false then
				local zero = 0
				db.query("INSERT INTO `player_stash` (`player_id`, `item_id`, `item_count`) VALUES (" .. JOGADOR:getGuid() .. ", " .. clientType:getClientId() .. ", " .. zero .. ");")
				end
					local itemCountPlayer = JOGADOR:getItemCount(item:getId())
					local resulttestId = db.storeQuery("SELECT `item_count` FROM `player_stash` WHERE `player_id` = " .. JOGADOR:getGuid() .. " AND `item_id` = " .. clientType:getClientId())
					local testando12 = result.getDataInt(resulttestId , "item_count")
					result.free(resulttestId)
					local TOTAL = math.floor(testando12 + itemCountPlayer)
					db.asyncQuery('UPDATE `player_stash` SET `item_count` = ' .. TOTAL .. ' WHERE `item_id` = '.. clientType:getClientId() ..' AND `player_id` = ' .. JOGADOR:getGuid())
					JOGADOR:removeItem(item:getId(), itemCountPlayer)
				end	
			end
		end	
	end
end
-- end of archive
```
As i said, only works if you stow you main backpack through the menu.
I know my scripts are bad, but i'm posting them here hoping that it can help someone.
The base script i got from 'quick loot' script!!

**Comment by vinioliveirasilva** [comment]
If someone could test it too, i will be thankful.

**Comment by marcosvf132** [comment]
> If someone could test it too, i will be thankful.

Is working fine here,
I have done a script to implement some minor features:
-stow specific item.
-stow main container.
-stow all items of same type.
Need to do:
-Make checks if have available depot slots on supply stash.
-stow specific container.
**-Most important: optimize the script.**

- On \otservbr-global\data\modules\scripts\supplystash\supplystash.lua

```lua

SupplyStash = {
	Size = 1000
}

SupplyStash.Credits = {
	Developer = "DudZ, Ticardo (Rick), gpedro, vinioliveirasilva",
	Version = "1.0",
	lastUpdate = "17/04/2020 - 12:00",
	todo = {"Missing create a custom stash depot", "Stow"}
}

SupplyStash.S_Packets= { OpenWindow = 0x29 }
SupplyStash.C_Packets = { Withdraw = 0x28 }

function onRecvbyte(player, msg, byte)
    if (byte == SupplyStash.C_Packets.Withdraw) then
		SupplyStash.parseStash(player, msg)
	else
		print("byte desconhecido - " ..byte)
    end
end

function loadStash(player, idToExcludeIfZeroCount, count)
	local items = {}
	local resultId = db.storeQuery("SELECT * FROM `player_stash` WHERE `player_id` = " .. player:getGuid())
	if resultId ~= false then
	    repeat
	    	itemId = result.getDataInt(resultId , "item_id")
	    	itemCount = result.getDataInt(resultId , "item_count")
	    	if (itemId ~= idToExcludeIfZeroCount) then 
    			items[itemId] = itemCount
			elseif (itemCount > count) then
				items[itemId] = itemCount - count
	    	end
		until not result.next(resultId)
	end	
	result.free(resultId)
	return items
end

function sendOpenWindowMessage(player, items)
	local count = 0
	local depotSlots = 1
	for _, itemCount in pairs(items) do
		depotSlots = depotSlots + math.ceil(math.max(itemCount / 100, 1))
		count = count + 1
	end
	local msg = NetworkMessage()
	msg:addByte(SupplyStash.S_Packets.OpenWindow);
	msg:addU16(count)
	for id, count in pairs(items) do
		msg:addU16(id)
		msg:addU32(count)
	end
	msg:addU16(SupplyStash.Size - depotSlots)
	msg:sendToPlayer(player)
	return true
end

SupplyStash.sendOpenWindow = function(player)
	local items = loadStash(player, 0, 0)
	return sendOpenWindowMessage(player, items)
end

SupplyStash.parseStash = function(player, msg)
	local action = msg:getByte()
	if action == 0 then -- STOW SPECIFIC ITEM
		local from1 = msg:getPosition() -- Item position.
		local itemCIDSTOW = msg:getU16() -- Item client id to be stow.
		local empty1 = msg:getByte() -- Unknown
		local itemCountSTOW = msg:getByte() -- Item count to be stow.
		if itemCIDSTOW and itemCountSTOW then if from1 then if from1.x == CONTAINER_POSITION then
		local container = player:getContainerById(from1.y - 64)
		if container then
		local item = container:getItem(from1.z)
		stowItemofThis(player, (Game.getItemIdByClientId(itemCIDSTOW)):getId(), itemCountSTOW)
		item:remove(itemCountSTOW)
		end	end	end	end
	elseif action == 1 then -- STOW CONTAINER
		local from = msg:getPosition() -- Picked container to stow.
		local test11 = msg:getByte() -- Unknown
		local test12 = msg:getByte() -- Unknown
		local test13 = msg:getByte() -- SLOT ID when stow from main backpack
		-- if from then
		-- if from.x == CONTAINER_POSITION then
			-- local container1 = player:getContainerById(from.y - 64)
			-- if container1 then
		-- local item2 = container1:getItem(from.z)
--			if item2:isContainer then
--			-- Add function to check on this container.
--			end
			-- end
		-- end		
		-- end
        searchitemsStash(player:getSlotItem(test13), player, true)		
	elseif action == 2 then -- STOW ALL ITEMS OF SAME TYPE
		local from = msg:getPosition() -- ??
		local itemClientId = msg:getU16() -- ??
		if itemClientId then
		local itemCount = player:getItemCount((Game.getItemIdByClientId(itemClientId)):getId())
		stowItemofThis(player, (Game.getItemIdByClientId(itemClientId)):getId(), itemCount)
		player:removeItem((Game.getItemIdByClientId(itemClientId)):getId(), itemCount)
		end
	elseif action == 3 then -- WITHDRAW ITEM
		local itemId = msg:getU16()
		local itemCount = msg:getU32()
		local items = loadStash(player, itemId, itemCount)
		if items[itemId] ~= nill then
			db.asyncQuery('UPDATE `player_stash` SET `item_count` = ' .. items[itemId] .. ' WHERE `item_id` = '.. itemId ..' AND `player_id` = ' .. player:getGuid())
		else
			db.asyncQuery('DELETE FROM `player_stash` WHERE `item_id` = ' .. itemId)
		end
		player:addItem((Game.getItemIdByClientId(itemId)):getId(), itemCount)
		return sendOpenWindowMessage(player, items)
	end
end
function searchitemsStash(test, JOGADOR, ALL)
	if test then
		for i = 0, test:getSize() do
				local item = test:getItem(i)
			if item then
				if (item:getCount() > 1) and not(item:getId() == 2160) and not(item:getId() == 2152) and not(item:getId() == 2148) and not(item:isContainer()) then
						local itemCountPlayer = JOGADOR:getItemCount(item:getId())
						if ALL then
						stowItemofThis(JOGADOR, item:getId(), itemCountPlayer)
						JOGADOR:removeItem(item:getId(), itemCountPlayer)
						else
						stowItemofThis(JOGADOR, item:getId(), item:getCount())
					    -- Remove only item count.
						end
				end	
					if item:isContainer() then
						searchitemsStash(item, JOGADOR, ALL)
					end
			end
		end
	else
	return false
	end
end

function stowItemofThis(JOGADOR1, ITEMID, COUNT)
  if ITEMID then
	local clientType = ItemType(ITEMID)
	local resulttestId = db.storeQuery("SELECT `item_count` FROM `player_stash` WHERE `player_id` = " .. JOGADOR1:getGuid() .. " AND `item_id` = " .. clientType:getClientId())
	result.free(resulttestId)
	  if resulttestId == false then
	   local zero = 0
	   db.query("INSERT INTO `player_stash` (`player_id`, `item_id`, `item_count`) VALUES (" .. JOGADOR1:getGuid() .. ", " .. clientType:getClientId() .. ", " .. zero .. ");")
	  end
	local resulttestId = db.storeQuery("SELECT `item_count` FROM `player_stash` WHERE `player_id` = " .. JOGADOR1:getGuid() .. " AND `item_id` = " .. clientType:getClientId())
	local testando12 = result.getDataInt(resulttestId , "item_count")
	result.free(resulttestId)
	db.asyncQuery('UPDATE `player_stash` SET `item_count` = ' .. (math.floor(testando12 + COUNT)) .. ' WHERE `item_id` = '.. clientType:getClientId() ..' AND `player_id` = ' .. JOGADOR1:getGuid())
	JOGADOR1:sendTextMessage(21, "Stowed " .. COUNT .. " objects.")
  end
end

```
Its possible to use the your 'loadstash(...)' function on the script to check if the player already have the item on his stash depot. maybe its a better way to make this script simple and short.

File: src/game.cpp [hunk 4]
```diff
@@ -1087,7 +1250,7 @@ void Game::playerMoveItem(Player* player, const Position& fromPos,
 	}
 	ReturnValue ret = internalMoveItem(fromCylinder, toCylinder, toIndex, item, count, nullptr, 0, player);
 	if (ret != RETURNVALUE_NOERROR) {
-		player->sendCancelMessage(ret);
+		return player->sendCancelMessage(ret);
```
Comment (gpedro) [inline src/game.cpp]: use return statement after method invocation cuz this method is void

File: src/game.cpp [hunk 5]
```diff
@@ -964,6 +995,137 @@ void Game::playerMoveItem(Player* player, const Position& fromPos,
...
+					if (itemDict[itemToRemove->getClientID()].first) {
+						internalRemoveItem(itemToRemove, itemToRemove->getItemCount());
+						totalStowed += itemToRemove->getItemCount();
+					}
+				}
+			}
+		}
+
+		query << "Stowed " << totalStowed << " object" << (totalStowed > 1 ? "s." : ".");
+		return player->sendCancelMessage(query.str());
```
Comment (gpedro) [inline src/game.cpp]: use return statement after method invocation cuz this method is void

File: src/game.cpp [hunk 6]
```diff
@@ -964,6 +995,137 @@ void Game::playerMoveItem(Player* player, const Position& fromPos,
 		}
 	}
 
+	//STASH
+	if (toCylinder->getContainer() != NULL &&
+		toCylinder->getItem()->getID() == ITEM_LOCKER1 &&
+		toPos.getZ() == ITEM_SUPPLY_STASH_INDEX) {
+
+		if (!isItemStorable(item)) {
+			return player->sendCancelMessage(RETURNVALUE_NOTPOSSIBLE);
```
Comment (gpedro) [inline src/game.cpp]: use return statement after method invocation cuz this method is void

File: src/game.cpp [hunk 7]
```diff
@@ -964,6 +995,137 @@ void Game::playerMoveItem(Player* player, const Position& fromPos,
...
+		std::ostringstream query;		
+
+		if (item->getContainer() != NULL) {
+			itemList = getAllStorableItemsInContainer(item);
+		} else {
+			itemList.push_back(item);
+		}
+
+		if (itemList.size() < 1) { //Safe check
+			return player->sendCancelMessage(RETURNVALUE_NOTPOSSIBLE);
```
Comment (gpedro) [inline src/game.cpp]: same here

**Comment by vinioliveirasilva** [comment]
Move communication data to cpp.

File: src/luascript.cpp [hunk 8]
```diff
@@ -11483,6 +11484,21 @@ int LuaScriptInterface::luaPlayerIsOffline(lua_State* L)
 	return 1;
 }
 
+int LuaScriptInterface::luaPlayerSetDepotStash(lua_State* L)
```
Comment (vinioliveirasilva) [inline src/luascript.cpp]: Remove

File: src/luascript.h [hunk 9]
```diff
@@ -1087,6 +1087,9 @@ class LuaScriptInterface
 		static int luaPlayerGetIdleTime(lua_State* L);
 		static int luaPlayerGetFreeBackpackSlots(lua_State* L);
 
+		// Stash
+		static int luaPlayerSetDepotStash(lua_State* L);
```
Comment (vinioliveirasilva) [inline src/luascript.h]: Remove

File: src/luascript.cpp [hunk 10]
```diff
@@ -2579,6 +2579,7 @@ void LuaScriptInterface::registerFunctions()
 	registerMethod("Player", "hasLearnedSpell", LuaScriptInterface::luaPlayerHasLearnedSpell);
 
 	registerMethod("Player", "sendImbuementPanel", LuaScriptInterface::luaPlayerSendImbuementPanel);
+	registerMethod("Player", "setDepotStash", LuaScriptInterface::luaPlayerSetDepotStash);
```
Comment (vinioliveirasilva) [inline src/luascript.cpp]: Remove

File: src/protocolgame.cpp [hunk 11]
```diff
@@ -1949,6 +1951,14 @@ void ProtocolGame::sendMarketEnter(uint32_t depotId)
 	sendResourcesBalance(player->getMoney(), player->getBankBalance());
 }
 
+void ProtocolGame::sendDepotStash(const bool enable)
```
Comment (vinioliveirasilva) [inline src/protocolgame.cpp]: Remove

File: src/protocolgame.h [hunk 12]
```diff
@@ -348,6 +348,9 @@ class ProtocolGame final : public Protocol
 		//shop
 		void AddShopItem(NetworkMessage& msg, const ShopInfo& item);
 
+		// stash supply
+		void sendDepotStash(bool enable);
```
Comment (vinioliveirasilva) [inline src/protocolgame.h]: Remove

**Comment by majestyotbr** [comment]
@gpedro @lgrossi PR waiting for your review.

File: src/actions.cpp [hunk 13]
```diff
@@ -545,8 +545,18 @@ bool Action::configureEvent(const pugi::xml_node& node)
 
 namespace {
 
-bool enterMarket(Player* player, Item*, const Position&, Thing*,
-														const Position&, bool) {
+bool enterStash(Player* player, Item*, const Position&, Thing*, const Position&, bool)
```
Comment (lgrossi) [inline src/actions.cpp:548]: Why keeping all those trash parameters if you only use player?
Please clean it up.

File: src/actions.cpp [hunk 14]
```diff
@@ -545,8 +545,18 @@ bool Action::configureEvent(const pugi::xml_node& node)
...
+{
+	if (player->getLastDepotId() == -1) {
+		return false;
+	}
+
+	player->sendOpenStash();
+	return true;
+}
+
+bool enterMarket(Player* player, Item*, const Position&, Thing*, const Position&, bool)
```
Comment (lgrossi) [inline src/actions.cpp:558]: Why keeping all those trash parameters if you only use player?
Please clean it up.

File: src/actions.cpp [hunk 15]
```diff
@@ -545,8 +545,18 @@ bool Action::configureEvent(const pugi::xml_node& node)
 
 namespace {
 
-bool enterMarket(Player* player, Item*, const Position&, Thing*,
-														const Position&, bool) {
+bool enterStash(Player* player, Item*, const Position&, Thing*, const Position&, bool)
+{
+	if (player->getLastDepotId() == -1) {
```
Comment (lgrossi) [inline src/actions.cpp:550]: what if player is a nullptr? please make it safer.

File: src/actions.cpp [hunk 16]
```diff
@@ -545,8 +545,18 @@ bool Action::configureEvent(const pugi::xml_node& node)
...
+		return false;
+	}
+
+	player->sendOpenStash();
+	return true;
+}
+
+bool enterMarket(Player* player, Item*, const Position&, Thing*, const Position&, bool)
+{
 	if (player->getLastDepotId() == -1) {
```
Comment (lgrossi) [inline src/actions.cpp:560]: what if player is a nullptr? please make it safer.

File: src/depotlocker.cpp [hunk 17]
```diff
@@ -24,7 +24,7 @@
 #include "depotlocker.h"
 
 DepotLocker::DepotLocker(uint16_t type) :
-	Container(type, 3), depotId(0) {}
+	Container(type, 4), depotId(0) {} // 4 so we have space for stash
```
Comment (lgrossi) [inline src/depotlocker.cpp:27]: why for? what is this number about? can it become more meaningful?

File: src/game.cpp [hunk 18]
```diff
@@ -1056,6 +1056,13 @@ void Game::playerMoveItem(Player* player, const Position& fromPos,
 		return;
 	}
 
+	if (toCylinder->getContainer() != NULL &&
+		toCylinder->getItem()->getID() == ITEM_LOCKER1 &&
```
Comment (lgrossi) [inline src/game.cpp:1060]: What if getItem() returns an empty pointer? please, make it safer.

File: src/game.cpp [hunk 19]
```diff
@@ -1056,6 +1056,13 @@ void Game::playerMoveItem(Player* player, const Position& fromPos,
...
+		toPos.getZ() == ITEM_SUPPLY_STASH_INDEX) {
+		player->stowContainer(item, count);
```
Comment (lgrossi) [inline src/game.cpp:1062]: is player validated anywhere? what if player is nullptr?

File: src/game.cpp [hunk 20]
```diff
@@ -1084,7 +1091,8 @@ void Game::playerMoveItem(Player* player, const Position& fromPos,
 	}
 	ReturnValue ret = internalMoveItem(fromCylinder, toCylinder, toIndex, item, count, nullptr, 0, player);
 	if (ret != RETURNVALUE_NOERROR) {
-		player->sendCancelMessage(ret);
+		 player->sendCancelMessage(ret);
```
Comment (lgrossi) [inline src/game.cpp:1094]: can't see any validation for player, what if it's a nullptr?

File: src/player.cpp [hunk 21]
```diff
@@ -4935,20 +4937,121 @@ void Player::onDeEquipImbueItem(Imbuement* imbuement)
...
+			std::cout << "[Player::addItemFromStash] Could not add itemId: " << itemId << " count: " << addValue << " to playerId: " << this->guid << std::endl;
+			delete newItem;
+			return false;
+		}
+	}
+	return true;
+}
+
+void Player::stowContainer(Item* item, uint32_t count) {
+	if (item == NULL || !isItemStorable(item)) {
```
Comment (lgrossi) [inline src/player.cpp:4958]: change null for nullptr.

File: src/player.cpp [hunk 22]
```diff
@@ -4935,20 +4937,121 @@ void Player::onDeEquipImbueItem(Imbuement* imbuement)
...
+    totalStowed += count;
+  }
+
+	retString << "Stowed " << totalStowed << " object" << (totalStowed > 1 ? "s." : ".");
+	sendCancelMessage(retString.str());
+}
+
+
+bool Player::isItemStorable(Item* item) {
+	auto isContainerAndHasSomethingInside = item->getContainer() != NULL && item->getContainer()->getItemList().size() > 0;
```
Comment (lgrossi) [inline src/player.cpp:5013]: what if item is nullptr?

File: src/player.cpp [hunk 23]
```diff
@@ -4935,20 +4937,121 @@ void Player::onDeEquipImbueItem(Imbuement* imbuement)
...
+	return (item->isStackable() &&
+		item->getID() != ITEM_GOLD_COIN &&
+		item->getID() != ITEM_PLATINUM_COIN &&
+		item->getID() != ITEM_CRYSTAL_COIN) ||
+		isContainerAndHasSomethingInside;
+}
+
+ItemDeque Player::getAllStorableItemsInContainer(Item* container) {
+
+	auto allITems = container->getContainer()->getItemList();
```
Comment (lgrossi) [inline src/player.cpp:5023]: what if container is nullptr?

File: src/protocolgame.cpp [hunk 24]
```diff
@@ -3905,25 +3907,73 @@ void ProtocolGame::sendItemsPrice()
...
-		return;
+  if (stackpos >= 10)
+    return;
 
+  NetworkMessage msg;
+
+  std::unordered_set<uint32_t>::iterator it = std::find(knownCreatureSet.begin(), knownCreatureSet.end(), creature->getID());
+  if (it != knownCreatureSet.end()) {
+    msg.addByte(0x6B);
+    msg.addPosition(creature->getPosition());
```
Comment (lgrossi) [inline src/protocolgame.cpp:3923]: what if creature is nullptr?

**Review by lgrossi — CHANGES_REQUESTED** [review]
a few comments about safeness and code standards, looks fine in general.

**Comment by Geekbuddy** [comment]
Tested. Works fine. Only that is missing is function for steward(hireling) to open the stash.

Image: https://geekbuddys.com/gallery/image/3-supply-stashpng/

**Comment by gpedro** [comment]
> Tested. Works fine. Only that is missing is function for steward(hireling) to open the stash.
> 
> Image: https://geekbuddys.com/gallery/image/3-supply-stashpng/

i found one missing feature and two bugs but I'm working on it!
- withdraw is using clientid instead serverid [FIXED]
- enable stow menu context [ADDED]
- stow context menu is only working for container (Stow All) [WORKING]

File: data/movements/scripts/others/tiles.lua [hunk 25]
```diff
@@ -35,6 +35,7 @@ function onStepIn(creature, item, position, fromPosition)
 			end
 
 			player:sendTextMessage(MESSAGE_STATUS_DEFAULT, "Your depot contains " .. depotItems .. " item" .. (depotItems > 1 and "s." or "."))
+			player:setDepotStash(true)
```
Comment (jlcvp) [inline data/movements/scripts/others/tiles.lua:38]: This attribute is only set when stepping on depot tile, If the player logs-out on it and then log back in this flag is not set and we got inconsistent behaviour

**Comment by majestyotbr** [comment]
@marcosvf132 did a new PR for Supply Stash. This can be closed.

**Pull request closed without merging**